### DISCUSSION TITLE
Configuration Loading Strategies

### DISCUSSION THREAD

**Discussion by kisa7** [nebulastream/nebulastream#190] [General]
This discussion introduces knowledge sharing in regard of how different industrial DBMSs manage their configurations and a mapping of these solutions to our project with potential advantages and disadvantages.

# Existing Solutions

As part of existing solutions, we investigated MongoDB, RockDB, ClickHouse, MySQL, and FoundationDB. Below is a description of how these systems handle configurations classified by centralized, hybrid, and decentralized solutions.

## Centralized

**MongoDB**
In terms of a single MongoDB instance, the configuration loading, parsing, and validation process are centralized within that instance before settings are applied to their respective components like storage, network, and others. 

1. Configuration Loading:
- When a MongoDB instance (mongod or mongos) starts up, it first loads the configuration from the specified sources. These can be a configuration file (typically YAML), command-line arguments, or environment variables.
2. Parsing and Validation:
- The loaded configurations are parsed and validated in a centralized manner. This means that all configurations are read and checked for correctness and validity (such as data types, range constraints, and dependencies among settings) before any operational components (like the network or storage engines) start up.
- This step ensures that only valid configurations are applied to the system components, preventing runtime failures due to misconfigurations.
3. Application of Settings:
- After validation, the settings are then distributed internally to the various components of the MongoDB instance:
4. Dynamic Reconfiguration:
- Some settings can be changed at runtime without restarting the instance. These changes are typically made via administrative commands that apply the new settings immediately to the relevant system component. Even these changes follow a centralized validation process within the instance before they are applied.

**RockDB**

1. Configuration Loading:
- When a RocksDB instance is initialized, it first loads configuration settings from the application code where the database is embedded. Configurations are typically set up through various C++ structs such as Options, DBOptions, and ColumnFamilyOptions.
- These configuration objects can be manually set in the code or loaded from an external source like a configuration file processed by the application, but RocksDB itself does not natively parse files or environment variables.
2. Parsing and Validation:
- Configuration objects like Options are filled with settings programmatically. Each setting is applied to the object properties directly in the code.
- As the configuration settings are applied, RocksDB internally validates these settings. For example, write_buffer_size must be within a practical and permissible range to ensure it's operational. 
3. Application of Settings:
- Once validated, the settings contained within these configuration objects are applied to the respective components of the RocksDB instance.
4. Dynamic Reconfiguration:
- Some options in RocksDB can be adjusted dynamically at runtime using methods like SetOptions(). This allows developers to modify certain parameters without needing to restart the database or application.
- Dynamic reconfigurations are validated in a similar manner to initial configurations to ensure they are within allowable ranges and do not conflict with other operational parameters.

**MYSQL**
1. Configuration Loading:
- When a MySQL instance (mysqld) starts up, it first loads the configuration from specified sources. These sources typically include configuration files (my.cnf or my.ini), which are the primary method for setting configuration options. Additionally, command-line arguments can be used to override settings in the configuration files, and in some cases, environment variables might influence settings.
- Configuration files can exist in multiple standard locations, and MySQL reads these files in a specific order, allowing settings in files read later to override those read earlier.
2. Parsing and Validation:
 - The loaded configurations are parsed in a centralized manner as MySQL starts. This process involves interpreting the settings specified in the configuration files and command-line arguments, and integrating these into the server's operational parameters.
- Each configuration setting is validated for correctness and applicability. This includes checking data types (ensuring string or numeric values are appropriate), range constraints (ensuring numeric values fall within acceptable limits), and dependencies among settings (e.g., buffer pool size relative to available system memory).
3. Application of Settings:
- After successful validation, the settings are applied internally to the various components of the MySQL instance.
4. Dynamic Reconfiguration:
- MySQL supports dynamic changes to certain settings while the server is running. These changes can be made through SQL commands, such as SET GLOBAL or SET SESSION, allowing administrators to adjust operational parameters without restarting the server.
- Dynamic changes are validated in the same thorough manner as at startup, ensuring that adjustments do not compromise the stability or security of the server.

## Hybrid

**ClickHouse**

ClickHouse uses XML files (config.xml and users.xml) for configuration, which is somewhat less common in modern DBMSs that often prefer simpler formats like YAML or JSON. XML allows for a more structured and hierarchical configuration, which can be particularly advantageous for complex setups with many nested options.

1. Configuration Loading:
- All configuration settings for ClickHouse are defined in XML files, such as config.xml for system settings and users.xml for user and access management settings. The configurations are stored in objects that can be accessed globally within the ClickHouse server process. 
- Each component within ClickHouse—whether it's the server itself, various storage engines, or network interfaces—accesses the configuration settings it requires from this centralized data structure as needed. This approach is somewhat like a "pull" model, where each component retrieves its specific configuration parameters.
- The components do not receive configurations passively; instead, they actively query the central configuration for the parameters relevant to their operational context.
2. Parsing and Validation:
- At startup, ClickHouse parses these XML files, loading settings into memory. During this initial load, some level of validation and parsing converts XML data into internal data structures that are more efficiently accessed by the server components.
3. Application of Settings:
- Server Configuration: Settings related to server operation, such as port numbers, memory limits, and path configurations, are applied to the server. This step configures how the server behaves in terms of network communication, data storage, and process management.
- Database and Table Settings: ClickHouse allows setting configurations at the database and table levels, influencing how data is stored, compressed, and accessed.
- User Profiles and Permissions: Configurations related to user profiles, access rights, and resource quotas are set up based on users.xml, defining how users interact with ClickHouse and what operations they are allowed to perform.
4. Dynamic Reconfiguration:
- For runtime modifications, certain settings can be dynamically changed through system management interfaces, such as SQL queries that adjust operational parameters. These changes are then reflected back into the central configuration state.
- Consistency and Coordination: For changes that affect the whole cluster, such as cluster-wide settings in a distributed ClickHouse setup, ClickHouse manages consistency by propagating changes across all nodes and ensuring that new settings do not disrupt ongoing operations.

## Decentralized 

**FoundationDB**

In terms of a single FoundationDB instance, the configuration process is decentralized and is designed to efficiently handle runtime changes across all client and server components. 

1. Configuration Loading:
- FoundationDB does not rely on a single central configuration file at startup. Instead, configurations can be specified through command-line options or set programmatically in the database itself.
2. Parsing and Validation:
- Instead of a single central validation step before starting operations, FoundationDB uses a global configuration framework. This framework broadcasts updates to configuration values, which are then independently managed and applied by each machine in the cluster. This setup allows for configurations to be adjusted on-the-fly without needing to restart processes or interfere with ongoing operations.
3.	Application of Settings:
- Configuration changes in FoundationDB are applied almost in real-time through the global configuration key space. Each component, whether client or server, checks this key space and applies updates to its operational parameters independently. This includes settings for network operations, data storage, and other critical functions.
- The system uses an eventually consistent model to ensure that all components across the cluster eventually synchronize on the configuration settings without necessitating simultaneous downtime or reboots.
4. Dynamic Reconfiguration:
- FoundationDB supports dynamic changes to configurations without service interruption. Changes are propagated through special transactions in the database and do not require direct file edits or command-line interventions post-initial setup.

# Potential Solutions

Overall, there are two ways how systems interpret centralized vs. decentralized: coordinator and workers level, and a worker and it's components level. We will discuss all of these options further.

## Centralized (Coordinator-Worker)

**SA1.** The coordinator stores all the configurations and propagates them to workers 
- **Disadvantage:** This approach can be inefficient as the coordinator can become a bottleneck in the case of thousands/millions of potential devices.
- **Disadvantage:** Assumes homogeneity of workers — in case different workers will require different configurations it will require providing a worker id for specific configurations. However that is not known before the worker launch.

**SA2.** The coordinator stores all the configurations, and workers pull necessary configurations as needed.
- **Advantage:** Ensures that all workers have access to the latest configurations without storing them locally, which can simplify updates and maintenance.
- **Disadvantage:** Assumes homogeneity of workers — in case different workers will require different configurations it will require providing a worker id for specific configurations. However that is not known before the worker launch.

_Enhancement for Both:_ Utilize ZooKeeper to store configurations, allowing for centralized management with decentralized access.

### Centralized (Worker-Components)

**SB1.** The worker stores all configurations, perform per-type validation and distributes them across different components (current implementation)
- **Advantage:** No redundant configurations
- **Advantage:** No redundant type-based (Int, Bool, String, Array, etc.) validations
- **Disadvantage:** No support for runtime reconfigurations. However, it could be implemented via query API
- **Disadvantage:** No semantic-based validation (for example, numberOfBuffersInSourceLocalBufferPool cannot be larger than available memory)

**SB2.** The worker stores all configurations, perform per-component validation and distributes them across different components 
- **Advantage:** Semantic based validation
- **Advantage:** No redundant type-based validations
- **Disadvantage:** No support for runtime reconfigurations. However, it could be implemented via query API
- **Disadvantage:** Duplicates in configurations options, as some configurations participate in multiple components (e.g. numberOfBuffersPerEpoch in fault-tolerance participates in the network stack, final sinks, query optimizer)

### Hybrid (Worker-Components)

**SC1.**  Worker stores all configurations, and components proactively pull configurations from a specialized structure that is created after validation of individual configuration options.
- **Advantage:** No redundant configurations 
- **Advantage:** No redundant type-based validations
- **Advantage:** Support of runtime reconfigurations
- **Disadvantage:** Imposes additional overhead on components, distracting them from their primary tasks and possibly affecting performance.
- **Disadvantage:** No semantic-based validation

### Decentralized (Worker-Components)

**SD1.** All components store and manage their own configurations independently.
- **Advantage:** Maximizes autonomy and responsiveness, as each component can quickly adapt configurations without dependencies.
- **Advantage:** Semantic-based validation
- **Disadvantage:** Risk of configuration drift where components become out of sync, leading to potential inconsistencies and duplication of configuration data across the system.
- **Disadvantage:** Potential redundant copies of configurations
- **Disadvantage:** Redundant typed-based validators. Could be potentially avoided if implemented in Util.
- **Disadvantage:** No support for runtime reconfigurations. Potential difficulties implementing it without pushing extra burdain on the user to be aware of existing system subcomponents.

**Comment by ls-1801** [comment]
I am not quite sure what you mean by store configuration in this context:

1. Does it refer to a `worker.yaml` that resides somewhere on the file system and is loaded by the worker?
2. Does it refer to the declaration of configurable parameters, as we do in the `WorkerConfiguration.hpp`?


Regarding the Centralized approach, maybe we should differentiate between Global, Local, and Query Parameters.

Global parameters should be synchronized between all workers and could thus be stored in the coordinator. I can't think of a global configuration right now. Maybe some restarting policies on worker crash.

Local parameters, would be something that is worker/machine specific, how much memory, threads, GPU...

Query Parameters: BufferSize, fault tolerance.

**Reply by kisa7** [reply]
Hey, thanks for pointing it out! I mean option 2!

Yes, I already thought of splitting configurations into two types: local and query. But I didn't think of the global ones! I also cannot come up with examples regarding global configuration from the point of existing configuration options. However this point can be an advantage for the centralized approach.

P.S. Fault tolerance has both types of configurations, local and query.

**Reply by keyseven123** [reply]
What is the difference for a configuration between a global and a local / query? 
Is the global a fallback, if the local or query is not set?

**Reply by kisa7** [reply]
No, these are just different types of configurations. For instance, a timezone could be a global configuration, as it is applied to all devices in the system. A numberOfBuffersInGlobalLocalBufferPool is a local configuration, as it sets the amount of memory allocated per worker. And a processingGuarantee is a fault tolerance guarantee set per query and specified as a part of the query.

**Comment by keyseven123** [comment]
>Disadvantage: This approach can be not efficient as all workers require configurations at launch, and waiting for the coordinator to send configurations can delay the startup processes.

But this should only be a problem once for each worker, right?

**Reply by kisa7** [reply]
This is generally the case. During the runtime reconfiguration, all workers are dependent on the coordinator propagating them changes. In the case of thousands of nodes, that might result in a delay.

Good point, I'll correct the Disadvantage.

**Comment by keyseven123** [comment]
> Disadvantage: Assumes homogeneity of workers — in case different workers will require different configurations it will require providing a worker id for specific configurations. However that is not known before the worker launch.

This is actually a really good point, IMO. With NebulaStream targeting hundreds or even thousands of nodes, this might be a lot of overhead. I would therefore argue that we should have as much as possible in a decentralized fashion and only go to a centralized if we absolute need to.

**Reply by kisa7** [reply]
Yes, I would also rather focus on the decentralized options.

**Comment by ls-1801** [comment]
I think the discussion is missing where and how the industrial DBMS declare their configuration parameters and how they are advertised to the system's user/developers. 

This only discusses where the configuration is loaded from, either via file, environment, or command line.

**Comment by zeuchste** [comment]
For SA1 and SA2: I think another downside is that you need for some paramters to have working defaults because you have to first start the engine before gathering happens. I am not sure if this is always possible.


I have a hard time deriving what configurations we are talking about here as for different types, there mayb be different methods benefical. My impression is there are:
 - Query related configuration, so things that only affect a single query
 - Engine related configuration, as the engine is our main working horse it may require an own handling like increasing number of worker threads
 - Debugging related configuration, like log level and dumping of information
 - User related configuration, like access control and setting configs for particular users
 - System-wide configuaration, so things that are set for all workers not for a particular one
 - Coordinator configuration, so things that only change the behavior of the coordinator
 - Network reladed configuration, 
 - And maybe many more

	In addition there are paramter that are static and will never be changed and some that might change during runtime

I think the exploration of the design space you did here is very good but we should also include the type of configuration here

**Comment by zeuchste** [comment]
In general if all solutions have pros and cons and we cannot really get one winner it is usually required that we define our own requirements and goals which is currently missing and see then which one is the best fit

**Comment by adrianmichalke** [comment]
My two cents. 
**1.** The disadvantages of the hybrid approach don't actually seem that heavy to me:
> Disadvantage: No semantic-based validation (for example, numberOfBuffersInSourceLocalBufferPool cannot be larger than available memory)

But this particular example can also be checked without the components having to interact.

> Disadvantage: Imposes additional overhead on components, distracting them from their primary tasks and possibly affecting performance.

I'm not sure if this is really such a problem. In the end, it's probably just a lookup in a global hash table.

**2.** I think we might want to have multiple options for loading configurations in our system:
1) Default configurations 
2) Set in a configuration yaml `worker.yml`
3) Set as command line argument `--host xxx`
4) Specified via [query API](https://duckdb.org/docs/configuration/overview.html)

This could also be the order in terms of configuration time by the user and how they override each other. Most of your proposed solutions would support such. 

For the default conf, we should strive for one that allows the user to try NES without configuring anything.

**3.** A very elegant way to implement runtime confs is to store them in a (virtual) table of the database itself. In the future this could also be an option for us. (which would be a centralized worker components approach)
